Maintenance -> Storage Migrations (#118)
polkadot-developers/polkadot-docs: develop/blockchains/maintenance/storage-migrations.md

- Before any transactions are processed

This timing is critical because the new runtime expects data to be in the updated format. Without proper migration, any attempt to decode the old data format could result in runtime panics or undefined behavior.

## When is a Storage Migration Required?

A storage migration is necessary whenever a runtime upgrade changes the storage layout or the encoding/interpretation of existing data. Even if the underlying data type appears to still "fit" the new storage representation, a migration may be required if the interpretation of the stored values has changed.

- Adding or removing an extrinsic introduces no new interpretation of preexisting data, so not migration required

- Reordering or mutating fields of an existing data type do change the encoded/decoded data representation, hence a storage migration is needed

- Removal of a pallet or storage item warrants cleaning up storage via a migration to avoid state bloat

- Adding a new storage item would not require any migration

### Common Cases

Here are some common scenarios where a storage migration is needed:

> - Changing data types:

Reviewer: if you are adding `:` at the end of this bullet point, ensure that the rest have them as well

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-1.rs'
    ```

>     Changing the underlying data type requires a migration to convert the existing values.

Reviewer: I'd dive deeper into the example to make it even more straightforward...

Thinking out loud:
"In this case, the example showcases a change from u32 data type to u64, however, this can happen with any allowed rust type."

- Changing data representation

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-2.rs'
    ```

    Modifying the representation of the stored data, even if the size appears unchanged, requires a migration to ensure the runtime can properly interpret the existing values.

- Extending an enum

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-3.rs'
    ```

>     Adding new variants to an enum requires a migration to handle the new variant, unless the new variant is added at the end and no existing values are initialized with it.

Reviewer: "the new variant is added at the end" seems confusing, at least for me. Could you please paraphrase it a bit?

- Changing the storage key

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-4.rs'
    ```

>     Modifying the storage key, even if the underlying data type remains the same, requires a migration to ensure the runtime can locate the correct stored values.

Reviewer: I think that in the future, we should explain storage for substrate chains. I believe that this part should link to that non-existing page since some devs might not know how the underlying storage mechanism in substrates works.

(however, I also think that if a dev gets to this point of doing a storage migration, they would have pretty solid knowledge about substrate storage, but it'd be worth it to have everything linked.)

Author: I agree. We should have a specif page explaining storage

Reviewer 2: FRAME storage has great API docs now: 

please contribute to it, when the time comes, and build on top of it with conscious repetition if needed. 

https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/attr.storage.html
same as with other macros: https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/index.html

!!!warning
    In general, any change to the storage layout or data encoding used in your runtime requires a careful consideration of the need for a storage migration. Overlooking a necessary migration can lead to undefined behavior or data loss during a runtime upgrade.

## Implementing Storage Migrations

The [`OnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.OnRuntimeUpgrade.html){target=\_blank} trait provides the foundation for implementing storage migrations in your runtime. Here's a detailed look at its key functions:

```rust
--8<-- 'code/develop/blockchains/maintenance/storage-migrations/on-runtime-upgrade-trait.rs'
```

### Core Migration Function

The **[`on_runtime_upgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.on_runtime_upgrade){target=\_blank}** function executes when the FRAME Executive pallet detects a runtime upgrade. Important considerations:

> - Runs before any pallet's `on_initialize` hooks

Reviewer: add link to that hook in the rust docs

> - Critical storage items (like block_number) may not be set

Reviewer: there should be a link to `block_number` in the system pallet i guess

- **[`post_upgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.post_upgrade){target=\_blank}** - after the migration completes, `post_upgrade` validates its success. It receives the state data captured by `pre_upgrade` and uses this to verify that the migration executed correctly. This function checks for storage consistency and ensures all data transformations occurred as expected. Like `pre_upgrade`, it operates exclusively in testing environments and should not modify storage

### Migration Logic

When implementing the actual migration logic, your code needs to handle several key responsibilities. Your migration implementation should:

- Read existing storage values in their original format
- Transform data to match the new format
- Write updated values back to storage
- Calculate and return consumed weight

> 

Reviewer: Add an action item that refers to testing

The recommended approach is to implement [`UncheckedOnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.UncheckedOnRuntimeUpgrade.html){target=\_blank} and wrap it with [`VersionedMigration`](https://paritytech.github.io/polkadot-sdk/master/frame_support/migrations/struct.VersionedMigration.html){target=\_blank}. `VersionedMigration` implements `OnRuntimeUpgrade` and handles storage version management automatically, following best practices and reducing potential errors.

`VersionedMigration` requires five type parameters:

- `From` - the source version for the upgrade
- `To` - the target version for the upgrade
- `Inner` - the `UncheckedOnRuntimeUpgrade` implementation
- `Pallet` - the pallet being upgraded
- `Weight` - the runtime's [`RuntimeDbWeight`](https://paritytech.github.io/polkadot-sdk/master/frame_support/weights/struct.RuntimeDbWeight.html){target=\_blank} implementation

```rust

> --8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/tags/polkadot-stable2409-1/substrate/frame/examples/single-block-migrations/src/migrations/v1.rs:0:122'

Reviewer: I know this might be a pain in the ass, but I think we should grab this remote file, taking only the lines we consider important. For example, all the comments at the very beginning are just making visual noise for somebody who reads this code snippet.

Reviewer: And also, somewhere a tiny paragraph should be added to explain what's exactly this coding is doing

```

### Migration Organization

Best practices recommend organizing migrations in a separate module within your pallet. Here's the recommended file structure:

```plain
my-pallet/
├── src/
│   ├── lib.rs       # Main pallet implementation
│   └── migrations/  # All migration-related code
│       ├── mod.rs   # Migrations module definition
│       ├── v1.rs    # V0 -> V1 migration
│       └── v2.rs    # V1 -> V2 migration
└── Cargo.toml
```

This structure provides several benefits:

- Separates migration logic from core pallet functionality
- Makes migrations easier to test and maintain
- Provides clear versioning of storage changes
- Simplifies the addition of future migrations

### Scheduling Migrations

To execute migrations during a runtime upgrade, you need to configure them in your runtime's `Executive` pallet. Add your migrations in `runtime/src/lib.rs`:

```rust

> --8<-- 'code/develop/blockchains/maintenance/storage-migrations/executive.rs'

Reviewer: I'd also dive a bit deeper into the comment "More migration can be added here", since it's a really important point of the whole migration workflow

```

### Single Block Migrations

Single block migrations execute their entire logic within one block immediately following a runtime upgrade. They run as part of the runtime upgrade process through the `OnRuntimeUpgrade` trait implementation and must complete before any other runtime logic executes.
While single block migrations are straightforward to implement and provide immediate data transformation, they carry significant risks. The most critical consideration is that they *must* complete within one block's weight limits. This is especially crucial for parachains, where exceeding block weight limits will brick the chain.
Use single block migrations only when you can guarantee:

- The migration has bounded execution time
- Weight calculations are thoroughly tested
- Total weight will never exceed block limits

For a complete implementation example of a single-block migration, refer to the [single block migration example]( https://paritytech.github.io/polkadot-sdk/master/pallet_example_single_block_migrations/index.html){target=\_blank} in the Polkadot SDK documentation.

> ### Multi Block Migrations

Reviewer: niceee, liked it 🚀

Reviewer: wdyt about moving "single block and multi block migrations" to be ## instead of ###

Or maybe they can be included in "## Migrations types" or whatever...

- After the new runtime is deployed
- Before any other code from the new runtime executes
- Before any [`on_initialize`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.on_initialize){target=\_blank} hooks run
- Before any transactions are processed

This timing is critical because the new runtime expects data to be in the updated format. Any attempt to decode the old data format without proper migration could result in runtime panics or undefined behavior.

## Storage Migration Scenarios

A storage migration is necessary whenever a runtime upgrade changes the storage layout or the encoding/interpretation of existing data. Even if the underlying data type appears to still "fit" the new storage representation, a migration may be required if the interpretation of the stored values has changed.

Storage migrations ensure data consistency and prevent corruption during runtime upgrades. Below are common scenarios categorized by their impact on storage and migration requirements:

> - Migration Required

Reviewer 3: ```suggestion
- Migration required:
```

    - Reordering or mutating fields of an existing data type to change the encoded/decoded data representation
    - Removal of a pallet or storage item warrants cleaning up storage via a migration to avoid state bloat

> - Migration Not Required

Reviewer 3: ```suggestion
- Migration not required:
```

- Before any transactions are processed

This timing is critical because the new runtime expects data to be in the updated format. Any attempt to decode the old data format without proper migration could result in runtime panics or undefined behavior.

## Storage Migration Scenarios

A storage migration is necessary whenever a runtime upgrade changes the storage layout or the encoding/interpretation of existing data. Even if the underlying data type appears to still "fit" the new storage representation, a migration may be required if the interpretation of the stored values has changed.

Storage migrations ensure data consistency and prevent corruption during runtime upgrades. Below are common scenarios categorized by their impact on storage and migration requirements:

=== "Migration Required"

    - Reordering or mutating fields of an existing data type to change the encoded/decoded data representation
    - Removal of a pallet or storage item warrants cleaning up storage via a migration to avoid state bloat

=== "Migration Not Required"

    - Adding a new storage item would not require any migration since no existing data needs transformation

>     - Adding or removing an extrinsic introduces no new interpretation of preexisting data, so no migration is required

Reviewer 3: In this case, where we’re explaining the reasoning behind why a migration is required or not, using tabs might imply that users need to choose between them, but they're not choosing, they're just learning at this point. Instead, presenting the options outside of tabs improves visibility and scan-ability, allowing users to easily see why a migration is needed or not.

Author: I've replaced this with nested list items. WDYT?

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-1.rs'
    ```

- **Changing data representation** - modifying the representation of the stored data, even if the size appears unchanged, requires a migration to ensure the runtime can correctly interpret the existing values

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-2.rs'
    ```

- **Extending an enum** - adding new variants to an enum requires a migration if you reorder existing variants, insert new variants between existing ones, or change the data type of existing variants. No migration is required when adding new variants at the end of the enum

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-3.rs'
    ```

- **Changing the storage key** - modifying the storage key, even if the underlying data type remains the same, requires a migration to ensure the runtime can locate the correct stored values.

    ```rust
    --8<-- 'code/develop/blockchains/maintenance/storage-migrations/example-4.rs'
    ```

!!!warning

>     In general, any change to the storage layout or data encoding used in your runtime requires carefulky considering the need for a storage migration. Overlooking a necessary migration can lead to undefined behavior or data loss during a runtime upgrade.

Reviewer 3: ```suggestion
    In general, any change to the storage layout or data encoding used in your runtime requires careful consideration of the need for a storage migration. Overlooking a necessary migration can lead to undefined behavior or data loss during a runtime upgrade.
```

## Implement Storage Migrations

The [`OnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.OnRuntimeUpgrade.html){target=\_blank} trait provides the foundation for implementing storage migrations in your runtime. Here's a detailed look at its essential functions:

```rust
--8<-- 'code/develop/blockchains/maintenance/storage-migrations/on-runtime-upgrade-trait.rs'
```

### Core Migration Function

> The **[`on_runtime_upgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.on_runtime_upgrade){target=\_blank}** function executes when the FRAME Executive pallet detects a runtime upgrade. Important considerations when using this function include:

Reviewer 3: ```suggestion
The [`on_runtime_upgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.on_runtime_upgrade){target=\_blank} function executes when the FRAME Executive pallet detects a runtime upgrade. Important considerations when using this function include:
```

The recommended approach is to implement [`UncheckedOnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.UncheckedOnRuntimeUpgrade.html){target=\_blank} and wrap it with [`VersionedMigration`](https://paritytech.github.io/polkadot-sdk/master/frame_support/migrations/struct.VersionedMigration.html){target=\_blank}. `VersionedMigration` implements `OnRuntimeUpgrade` and handles storage version management automatically, following best practices and reducing potential errors.

`VersionedMigration` requires five type parameters:

- `From` - the source version for the upgrade
- `To` - the target version for the upgrade
- `Inner` - the `UncheckedOnRuntimeUpgrade` implementation
- `Pallet` - the pallet being upgraded
- `Weight` - the runtime's [`RuntimeDbWeight`](https://paritytech.github.io/polkadot-sdk/master/frame_support/weights/struct.RuntimeDbWeight.html){target=\_blank} implementation

> Let's examine a migration example that transforms a simple `StorageValue` storing a `u32` into a more complex structure that tracks both current and previous values using the `CurrentAndPreviousValue` struct:

Reviewer 3: ```suggestion
Examine the following migration example that transforms a simple `StorageValue` storing a `u32` into a more complex structure that tracks both current and previous values using the `CurrentAndPreviousValue` struct:
```

Avoid let's

- Old `StorageValue` format:
```rust
#[pallet::storage]
pub type Value<T: Config> = StorageValue<_, u32>;
```

- New `StorageValue` format:
```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/tags/polkadot-stable2409-1/substrate/frame/examples/single-block-migrations/src/lib.rs:166:177'

--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/tags/polkadot-stable2409-1/substrate/frame/examples/single-block-migrations/src/lib.rs:200:201'
```

> - Migration

Reviewer 3: ```suggestion
- Migration:
```

```rust
--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/tags/polkadot-stable2409-1/substrate/frame/examples/single-block-migrations/src/migrations/v1.rs:18:122'
```

### Migration Organization

Best practices recommend organizing migrations in a separate module within your pallet. Here's the recommended file structure:

```plain
my-pallet/
├── src/
│   ├── lib.rs       # Main pallet implementation
│   └── migrations/  # All migration-related code
│       ├── mod.rs   # Migrations module definition
│       ├── v1.rs    # V0 -> V1 migration
│       └── v2.rs    # V1 -> V2 migration
└── Cargo.toml
```

This structure provides several benefits:

- Separates migration logic from core pallet functionality
- Makes migrations easier to test and maintain
- Provides explicit versioning of storage changes
- Simplifies the addition of future migrations

### Scheduling Migrations

> To execute migrations during a runtime upgrade, you must configure them in your runtime's `Executive` pallet. Add your migrations in `runtime/src/lib.rs`:

Reviewer 3: ```suggestion
To execute migrations during a runtime upgrade, you must configure them in your runtime's Executive pallet. Add your migrations in `runtime/src/lib.rs`:
```

The recommended approach is to implement [`UncheckedOnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.UncheckedOnRuntimeUpgrade.html){target=\_blank} and wrap it with [`VersionedMigration`](https://paritytech.github.io/polkadot-sdk/master/frame_support/migrations/struct.VersionedMigration.html){target=\_blank}. `VersionedMigration` implements `OnRuntimeUpgrade` and handles storage version management automatically, following best practices and reducing potential errors.

`VersionedMigration` requires five type parameters:

- `From` - the source version for the upgrade
- `To` - the target version for the upgrade
- `Inner` - the `UncheckedOnRuntimeUpgrade` implementation
- `Pallet` - the pallet being upgraded
- `Weight` - the runtime's [`RuntimeDbWeight`](https://paritytech.github.io/polkadot-sdk/master/frame_support/weights/struct.RuntimeDbWeight.html){target=\_blank} implementation

```rust

> use frame_support::{

Author: We can discuss this further, but the goal of the Develop section is to provide an overview of what developers can accomplish and show general guidance. Detailed, step-by-step examples would be included in the Tutorials section, with links from this page. What do you think?
